src/libAtomVM/nifs.c

```diff
@@ -1460,6 +1482,76 @@ static term nifs_erlang_process_info(Context *ctx, int argc, term argv[])
     return ret;
 }
 
+static int set_system_architecture(char *buf, size_t max_len)
+{
+    size_t system_name_len = strlen(SYSTEM_NAME);
+    size_t system_version_len = strlen(SYSTEM_VERSION);
+    size_t system_architecture_len = strlen(SYSTEM_ARCHITECTURE);
+    if ((system_name_len + 1 + system_version_len + 1 + system_architecture_len + 1) > max_len) {
+        return -1;
+    }
+    char *p = buf;
+    memcpy(p, SYSTEM_NAME, system_name_len);
+    p += system_name_len;
+    p[0] = '-';
+    p += 1;
+    memcpy(p, SYSTEM_VERSION, system_version_len);
+    p += system_version_len;
+    p[0] = '-';
+    p += 1;
+    memcpy(p, SYSTEM_ARCHITECTURE, system_architecture_len);
+    p += system_architecture_len;
+    p[0] = '\0';
+    return 0;
+}
+
+static size_t binary_wordsize(size_t len)
+{
+    return 2 + len / TERM_BYTES + (len % TERM_BYTES ? 1 : 0);
+}
+
+static term nifs_erlang_system_info(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term key = argv[0];
+    
+    if (!term_is_atom(key)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    
+    if (key == PROCESS_COUNT_ATOM) {
+        return term_from_int32(nifs_num_processes(ctx->global));
+    }
+    if (key == PORT_COUNT_ATOM) {
+        return term_from_int32(nifs_num_ports(ctx->global));
+    }
+    if (key == ATOM_COUNT_ATOM) {
+        return term_from_int32(ctx->global->atoms_table->count);
+    }
+    if (key == WORDSIZE_ATOM) {
+        return term_from_int32(TERM_BYTES);
+    }
+    if (key == VERSION_ATOM) {
+        size_t len = strlen(GIT_REVISION);
+        if (memory_ensure_free(ctx, binary_wordsize(len)) != MEMORY_GC_OK) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        return term_from_literal_binary((const uint8_t *) GIT_REVISION, len, ctx);
+    }
+    if (key == SYSTEM_ARCHITECTURE_ATOM) {
+        char buf[128];
+        if (set_system_architecture(buf, 128) == -1) {
+            RAISE_ERROR(SYSTEM_LIMIT_ATOM);
+        }
+        size_t len = strlen(buf);
+        if (memory_ensure_free(ctx, binary_wordsize(len)) != MEMORY_GC_OK) {
```

Comment:
there are some examples around about how to ensure memory for binaries, take a look to them.

Reply:
Using the relatively recently added term_binary_data_size_in_terms